## Regression Modeling and Evaluation

This section represents the third component of the mentorship project, focusing on building and evaluating multiple regression models to predict taxi fare amounts. The preprocessed and scaled dataset will be used to train and compare the performance of the following algorithms:

- **Linear Regression**
- **Decision Tree Regressor**
- **Random Forest Regressor**
- **Support Vector Regressor (SVR)**
- **Extreme Gradient Boosting Regressor(XGboost)**

### Model Evaluation Metrics
Each model will be evaluated using the following metrics to assess prediction accuracy and generalization:

- **Mean Squared Error (MSE)**
- **Root Mean Squared Error (RMSE)**
- **R² Score**
- **Mean Absolute Error(MAE)**

### Hyperparameter Tuning
To enhance model performance, we will perform hyperparameter tuning using:

- **Grid Search CV** (Exhaustive parameter search)
- **Randomized Search CV** (Efficient randomized parameter sampling)

The best model will be selected based on cross-validated performance metrics and will be considered for deployment or further integration.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, RandomizedSearchCV, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import StandardScaler
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, root_mean_squared_error
%matplotlib inline

In [2]:
#from component 2...
clean_df = pd.read_csv("clean_data.csv")
X = clean_df.iloc[:, :-1].values
y = clean_df.iloc[:, -1].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#scaling the data
sc = StandardScaler()
X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

### **Now in this 3rd component's 1st task we will be Building a baseline model with the selected features without any preprocessing, just to get a baseline understanding of how our models perform on raw data. next we will use the feature engineered and scaled  data from the component 2 and again implement the Machine learning algorithms and see how the accuracy spiked up or down.**

In [3]:
clean_df.columns

Index(['Unnamed: 0', 'distance_km', 'fare_per_km', 'day_of_week', 'year',
       'hour', 'fare_amount'],
      dtype='object')

In [4]:
# implementing Multiple Linear regression model as we have more then 1 features

MLR = LinearRegression()
MLR.fit(X_train, y_train)
y_hat = MLR.predict(X_test)

print(f"the MSE of the MLR model is {mean_squared_error(y_test, y_hat)}")
print(f"the RMSE of the MLR model is {root_mean_squared_error(y_test, y_hat)}")
print(f"the MAE of the MLR model is {mean_absolute_error(y_test, y_hat)}")
print(f"the R2 score of the MLR model is {r2_score(y_test, y_hat) * 100}%")

the MSE of the MLR model is 88.53686583354587
the RMSE of the MLR model is 9.409403054048958
the MAE of the MLR model is 5.858866872366144
the R2 score of the MLR model is 4.208019254766238%


In [5]:
RF = RandomForestRegressor(n_estimators=100)
RF.fit(X_train, y_train)
y_hat = RF.predict(X_test)

print(f"the MSE of the RandomForestRegressor model is {mean_squared_error(y_test, y_hat)}")
print(f"the RMSE of the RandomForestRegressor model is {root_mean_squared_error(y_test, y_hat)}")
print(f"the MAE of the RandomForestRegressor model is {mean_absolute_error(y_test, y_hat)}")
print(f"the R2 score of the RandomForestRegressor model is {r2_score(y_test, y_hat) * 100}%")

the MSE of the RandomForestRegressor model is 0.820380213280598
the RMSE of the RandomForestRegressor model is 0.9057484271477362
the MAE of the RandomForestRegressor model is 0.04907794588229421
the R2 score of the RandomForestRegressor model is 99.11239408742917%


In [6]:
# lets train and see how decision tree performs on baseline predection
DT = DecisionTreeRegressor()
DT.fit(X_train, y_train)
y_hat = DT.predict(X_test)

print(f"the MSE of the DecisionTreeRegressor model is {mean_squared_error(y_test, y_hat)}")
print(f"the RMSE of the DecisionTreeRegressor model is {root_mean_squared_error(y_test, y_hat)}")
print(f"the MAE of the DecisionTreeRegressor model is {mean_absolute_error(y_test, y_hat)}")
print(f"the R2 score of the DecisionTreeRegressor model is {r2_score(y_test, y_hat) * 100}%")

the MSE of the DecisionTreeRegressor model is 0.9061721449965442
the RMSE of the DecisionTreeRegressor model is 0.951930745903579
the MAE of the DecisionTreeRegressor model is 0.08354837058085852
the R2 score of the DecisionTreeRegressor model is 99.01957197323235%


In [7]:
xgb = XGBRegressor()
xgb.fit(X_train, y_train)
y_hat = xgb.predict(X_test)

print(f"the MSE of the extreme gradient boosting(XGB) regressor model is {mean_squared_error(y_test, y_hat)}")
print(f"the RMSE of the extreme gradient boosting(XGB) regressor model is {root_mean_squared_error(y_test, y_hat)}")
print(f"the MAE of the extreme gradient boosting(XGB) regressor model is {mean_absolute_error(y_test, y_hat)}")
print(f"the R2 score of the extreme gradient boosting(XGB) regressor model is {r2_score(y_test, y_hat) * 100}%")

the MSE of the extreme gradient boosting(XGB) regressor model is 5.777601179954096
the RMSE of the extreme gradient boosting(XGB) regressor model is 2.4036641154608303
the MAE of the extreme gradient boosting(XGB) regressor model is 0.3568002053705182
the R2 score of the extreme gradient boosting(XGB) regressor model is 93.7489558075806%


### Model Performance (Baseline - Unscaled)

| Model                        | MSE      | RMSE     | MAE      | R² Score (%)   |
|-----------------------------|----------|----------|----------|----------------|
| Linear Regression (MLR)     | 88.54    | 9.41     | 5.86     | 4.21%          |
| Random Forest Regressor     | 0.82     | 0.91     | 0.049    | 99.11%         |
| Decision Tree Regressor     | 0.91     | 0.95     | 0.084    | 99.02%         |
| XGBoost Regressor           | 5.78     | 2.40     | 0.357    | 93.75%         |

### Observations

- The **Multiple Linear Regression model** performed poorly with a very low R² score (4.2%), indicating that it struggles to capture the underlying patterns in the raw dataset.
- In contrast, **tree-based models (Random Forest, Decision Tree, XGBoost)** performed exceptionally well, with R² scores above 93%, especially Random Forest with **>99%** accuracy.
- These results suggest the models might be **overfitting** the data due to the absence of scaling or other regularization techniques.
- To verify and validate these high scores, further modeling steps will include **data scaling, cross-validation, and hyperparameter tuning**.

 Next, I will evaluate these models using **scaled data** to examine whether preprocessing improves performance or reduces overfitting risk.


In [ ]:
# now as we usually need to train models we will create a re-usable function to train models
def run_model_evaluation(model, x_t, y_t, x_test=X_test_scaled):
  model = model.lower()
  if model == "randomforest":
    model = RandomForestRegressor(n_estimators=100)
  elif model == "decisiontree":
    model = DecisionTreeRegressor()
  elif model == "xgb":
    model = XGBRegressor()
  elif model == "linearregression":
    model = LinearRegression()
  else:
    raise ValueError("you haven't passed a valid model ensure it's in the given list [randomforest,decisiontree,xgb,linearregression]")
  model.fit(x_t, y_t)
  y_hat = model.predict(x_test)

  mse = mean_squared_error(y_test, y_hat)
  mae = mean_absolute_error(y_test, y_hat)
  # rmse = root_mean_squared_error(y_test, y_hat) "or we can just sq root the mse"
  rmse = np.sqrt(mse)
  r2 = r2_score(y_test, y_hat)

  results = pd.DataFrame([{
    "Model": model.__class__.__name__,
    "MSE": mse,
    "MAE": mae,
    "RMSE": rmse,
    "R2 Score": r2
  }])
  return results


In [9]:
#using the scaled data with the selected features
models = ["LinearRegression", "DecisionTree", "XGB", "RandomForest"]
for model in models:
  print(run_model_evaluation(model, x_t=X_train_scaled, y_t=y_train))

              Model        MSE       MAE      RMSE  R2 Score
0  LinearRegression  88.536866  5.858867  9.409403   0.04208
                   Model       MSE       MAE      RMSE  R2 Score
0  DecisionTreeRegressor  0.986356  0.083076  0.993154  0.989328
          Model       MSE      MAE      RMSE  R2 Score
0  XGBRegressor  6.719739  0.36693  2.592246  0.927296
                   Model       MSE       MAE      RMSE  R2 Score
0  RandomForestRegressor  0.804586  0.048415  0.896987  0.991295


## Regression Models with Scaled Data

After establishing a baseline, I proceeded to scale the feature-engineered dataset to normalize the value ranges and improve model generalization. Standardization helps certain algorithms (especially those sensitive to feature magnitudes like Linear Regression) perform better and can also reduce variance in tree-based models.

### Model Performance (With Scaled Data)

| Model                  | MSE       | RMSE     | MAE      | R² Score (%)   |
|------------------------|-----------|----------|----------|----------------|
| Linear Regression      | 88.54     | 9.41     | 5.86     | 4.21%          |
| Decision Tree Regressor| 0.99      | 0.99     | 0.083    | 98.93%         |
| Random Forest Regressor| 0.80      | 0.90     | 0.048    | 99.13%         |
| XGBoost Regressor      | 6.72      | 2.59     | 0.367    | 92.73%         |

### Observations

- **Linear Regression** again showed poor performance with low R², confirming that linear assumptions do not capture the data patterns effectively, even after scaling.
- **Tree-based models** maintained their superior performance even with scaled features, showing consistent R² scores near or above 99%.
- The **Random Forest Regressor** slightly improved with scaling and remains the best performer in terms of both error metrics and R² score.
- **XGBoost** also performed well, though slightly lower compared to its baseline run, suggesting minor sensitivity to scaled data.

To further ensure that these excellent results are not due to chance or overfitting, the next step involves **cross-validating** the models using k-fold cross-validation.


In [10]:
def run_cross_validation(model):
  scores = cross_val_score(model, X_train_scaled, y_train, cv =5, scoring="r2", n_jobs=-1)
  print(f"The cross validated scores are {scores}")
  print(f"the score is {np.mean(scores)}")
  print(f"The standerd deviation is {np.std(scores)}")

In [11]:
run_cross_validation(RandomForestRegressor(n_estimators=100, random_state=42))

The cross validated scores are [0.99464011 0.95243179 0.98868618 0.99499571 0.99167634]
the score is 0.9844860270316633
The standerd deviation is 0.01618873003060893


In [13]:
run_cross_validation(XGBRegressor(tree_method="hist", device="cuda"))

The cross validated scores are [0.92453265 0.85797811 0.9224945  0.94141019 0.9121489 ]
the score is 0.9117128709564589
The standerd deviation is 0.028460975296510024


In [14]:
run_cross_validation(DecisionTreeRegressor())

The cross validated scores are [0.9909752  0.94114308 0.98312335 0.98598108 0.98673354]
the score is 0.9775912520534357
The standerd deviation is 0.018396597846566266


## Cross-Validation to Assess Model Stability

To ensure the high performance of the models wasn't due to overfitting or lucky data splits, I applied **5-Fold Cross-Validation** on the top-performing scaled models: **Random Forest**, **XGBoost**, and **Decision Tree**. This process evaluates model stability and generalization across different data subsets.

###  Cross-Validation Results (R² Scores)

| Model                  | Fold Scores                                           | Mean R² (%)   | Std Dev      |
|------------------------|------------------------------------------------------|---------------|--------------|
| Random Forest Regressor| [0.9946, 0.9524, 0.9887, 0.9950, 0.9917]              | **98.45%**    | 0.0162       |
| Decision Tree Regressor| [0.9910, 0.9411, 0.9831, 0.9860, 0.9867]              | **97.76%**    | 0.0184       |
| XGBoost Regressor      | [0.9245, 0.8580, 0.9225, 0.9414, 0.9121]              | **91.17%**    | 0.0285       |

### Observations

- **Random Forest Regressor** remains the most robust model, achieving the highest mean R² and lowest standard deviation, indicating strong generalization.
- **Decision Tree Regressor** also performs very well with low variance and high R².
- **XGBoost**, while slightly lower in performance, is still reliable and stable with a decent mean R² score and moderate variability.
- **Linear Regression** was excluded from cross-validation due to its consistently poor performance in earlier experiments.

With consistent results across folds, we can now confidently move forward with **hyperparameter tuning** to squeeze out further performance gains.


In [15]:
# lets tune the model with hyperparameter tuning we will create a re-usable function to tune the hyperparameters.
def tune_hyperparameter(model, param_grid, X_train, y_train, cv=3, scoring="r2", n_jobs = -1):
  randomized_search = RandomizedSearchCV(estimator=model, param_distributions=param_grid, scoring=scoring, n_jobs=n_jobs)
  randomized_search.fit(X_train, y_train)
  scores = {
      "best_parameters": randomized_search.best_params_,
      "best_score": randomized_search.best_score_
  }
  return scores

In [16]:
param_grids_and_models = {
      'XGBregressor': (XGBRegressor(tree_method="hist", device="cuda"), {
          'n_estimators': [100, 200, 300],
          'max_depth': [3, 6, 10],
          'learning_rate': [0.01, 0.1, 0.2],
          'subsample': [0.7, 0.8, 1.0],
          'colsample_bytree': [0.7, 0.8, 1.0],
          'gamma': [0, 0.1, 0.2],
          'booster': ['gbtree', 'dart']
    }),'DecisionTreeRegressor': (DecisionTreeRegressor(), {
        'max_depth': [10, 20, 30, None],
        'min_samples_split': [2, 5, 10],
        'min_samples_leaf': [1, 2, 4],
        'max_features': ['sqrt', 'log2', None],
        'splitter': ['best', 'random'],
        'criterion': ['squared_error', 'friedman_mse', 'absolute_error', 'poisson']
})
}


In [17]:
#for XGBoost and Decision tree **we are not including the Random forest model cause it takes lots of resources insted we will tune the two models and check the scores if the two models outperformed Random forest cross validates score we will use them as final model if not we will train final model as  random forest. **
for key, (model, param) in param_grids_and_models.items():
  print(f"Tuning {key}...")
  scores = tune_hyperparameter(model=model, param_grid=param, X_train=X_train_scaled, y_train=y_train)
  print(scores)

Tuning XGBregressor...


/usr/local/lib/python3.11/dist-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


{'best_parameters': {'subsample': 0.8, 'n_estimators': 100, 'max_depth': 6, 'learning_rate': 0.2, 'gamma': 0.2, 'colsample_bytree': 0.7, 'booster': 'gbtree'}, 'best_score': np.float64(0.9207319531340261)}
Tuning DecisionTreeRegressor...
{'best_parameters': {'splitter': 'random', 'min_samples_split': 5, 'min_samples_leaf': 1, 'max_features': None, 'max_depth': 30, 'criterion': 'absolute_error'}, 'best_score': np.float64(0.940474006558255)}


## Hyperparameter Tuning & Final Model Selection

To improve performance, I used **RandomizedSearchCV** with **3-fold cross-validation** for hyperparameter tuning on selected models.

### Models Tuned:
- **XGBoost Regressor**
- **Decision Tree Regressor**

> **Note:** Although Random Forest was excluded from tuning due to resource constraints, it still outperformed all other models **even without hyperparameter tuning** — and hence, it was chosen as the final model.


In [23]:
X = clean_df.iloc[:, :-1].values
y = clean_df.iloc[:, -1].values
scaler = StandardScaler()
X = scaler.fit_transform(X)
final_regressor = RandomForestRegressor(n_estimators=100, random_state=42)
final_regressor.fit(X, y)

RandomForestRegressor(random_state=42)

In [24]:
clean_df.columns

Index(['Unnamed: 0', 'distance_km', 'fare_per_km', 'day_of_week', 'year',
       'hour', 'fare_amount'],
      dtype='object')

## Final Model Selection & Training

After evaluating multiple regression models and their performance using both raw and scaled data, cross-validation, and hyperparameter tuning, I selected the **Random Forest Regressor** as the final model.

### Why Random Forest?
- Achieved **highest R² score (0.991)** even without tuning.
- Maintained **stable and consistent performance** during cross-validation.
- Demonstrated **low error metrics** (MSE, RMSE, MAE).
- Handles non-linear relationships and feature interactions very well.

### Final Step:
The final model was **trained on the entire scaled dataset** to fully utilize all available information.


In [26]:
clean_df.head()

,Unnamed: 0,distance_km,fare_per_km,day_of_week,year,hour,fare_amount
0,0,1.683325,4.452822,3,2015,19,7.5
1,1,2.457593,3.131872,4,2009,20,7.7
2,2,5.036384,2.560853,0,2009,21,12.9
3,3,1.661686,3.187614,4,2009,8,5.3
4,4,4.475456,3.574256,3,2014,17,16.0


In [28]:
# Feature order: ['distance_km', 'fare_per_km', 'day_of_week', 'year', 'hour']
prediction = final_regressor.predict([[0, 7.5, 3.2, 2, 2023, 18]]) # Example: 7.5 km, ₹3.2/km, Tuesday, 2023, 6 PM

print(f"the predection of the fair for the given input is: ${prediction[0]}")

the predection of the fair for the given input is: $198.5197


## 📌 Final Remarks

In this project, i conducted a comprehensive analysis of Uber fare prediction using regression modeling. The end-to-end workflow included:

- Exploratory Data Analysis (EDA)
- Feature Engineering and Preprocessing
- Baseline Model Evaluation
- Model Training with Scaled Data
- Performance Comparison of Multiple Regressors:
  - Linear Regression
  - Decision Tree Regressor
  - Random Forest Regressor
  - XGBoost Regressor
- Cross-Validation to ensure model generalization
- Hyperparameter Tuning using `RandomizedSearchCV`
- Final Model Selection and Training (Random Forest Regressor)

Each step was designed to ensure no data leakage, avoid overfitting, and maintain a balance between accuracy and interpretability. The final model was trained on the full preprocessed dataset, ensuring optimal learning from all available information.

---

### ✍️ Author

**Ankit Chimariya**  
*Data Science Intern*  
📧 chimariyaankit234@email.com<br>
📅 *Project Completed on: May 2025*

---
